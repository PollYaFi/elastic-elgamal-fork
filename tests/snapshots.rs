//! Snapshot testing to check compatibility of `serde` serialization of types.

use insta::assert_yaml_snapshot;
use rand_chacha::ChaChaRng;
use rand_core::SeedableRng;

use elastic_elgamal::{
    app::{ChoiceParams, EncryptedChoice},
    group::{Generic, Group, Ristretto},
    Keypair, RangeDecomposition,
};

trait Named {
    const NAME: &'static str;
}

impl Named for Ristretto {
    const NAME: &'static str = "ristretto";
}

impl Named for Generic<::k256::Secp256k1> {
    const NAME: &'static str = "k256";
}

fn test_ciphertext_snapshot<G: Group + Named>() {
    let mut rng = ChaChaRng::seed_from_u64(12345);
    let (public_key, _) = Keypair::<G>::generate(&mut rng).into_tuple();
    let ciphertext = public_key.encrypt(42, &mut rng);

    let full_name = format!("ciphertext-{}", G::NAME);
    assert_yaml_snapshot!(full_name, ciphertext);
}

fn test_zero_encryption_snapshot<G: Group + Named>() {
    let mut rng = ChaChaRng::seed_from_u64(12345);
    let (public_key, _) = Keypair::<G>::generate(&mut rng).into_tuple();

    let (ciphertext, proof) = public_key.encrypt_zero(&mut rng);
    let ciphertext_with_proof = serde_json::json!({
        "ciphertext": ciphertext,
        "proof": proof,
    });
    let full_name = format!("zero-encryption-{}", G::NAME);
    assert_yaml_snapshot!(full_name, ciphertext_with_proof);
}

fn test_bool_encryption_snapshot<G: Group + Named>() {
    let mut rng = ChaChaRng::seed_from_u64(12345);
    let (public_key, _) = Keypair::<G>::generate(&mut rng).into_tuple();

    let (ciphertext, proof) = public_key.encrypt_bool(true, &mut rng);
    let ciphertext_with_proof = serde_json::json!({
        "ciphertext": ciphertext,
        "proof": proof,
    });
    let full_name = format!("bool-encryption-{}", G::NAME);
    assert_yaml_snapshot!(full_name, ciphertext_with_proof);
}

fn test_range_encryption_snapshot<G: Group + Named>() {
    let mut rng = ChaChaRng::seed_from_u64(12345);
    let (public_key, _) = Keypair::<G>::generate(&mut rng).into_tuple();
    let range = RangeDecomposition::optimal(100).into();

    let (ciphertext, proof) = public_key.encrypt_range(&range, 42, &mut rng);
    let ciphertext_with_proof = serde_json::json!({
        "ciphertext": ciphertext,
        "proof": proof,
    });
    let full_name = format!("range-encryption-{}", G::NAME);
    assert_yaml_snapshot!(full_name, ciphertext_with_proof);
}

fn test_encrypted_choice_snapshot<G: Group + Named>() {
    let mut rng = ChaChaRng::seed_from_u64(12345);
    let (public_key, _) = Keypair::<G>::generate(&mut rng).into_tuple();
    let choice_params = ChoiceParams::single(public_key, 5);

    let choice = EncryptedChoice::single(3, &choice_params, &mut rng);
    let full_name = format!("encrypted-choice-{}", G::NAME);
    assert_yaml_snapshot!(full_name, choice);
}

fn test_encrypted_multi_choice_snapshot<G: Group + Named>() {
    let mut rng = ChaChaRng::seed_from_u64(12345);
    let (public_key, _) = Keypair::<G>::generate(&mut rng).into_tuple();
    let choice_params = ChoiceParams::multi(public_key, 5);

    let choices = [false, true, true, false, true];
    let choices = EncryptedChoice::new(&choice_params, &choices, &mut rng);
    let full_name = format!("encrypted-multi-choice-{}", G::NAME);
    assert_yaml_snapshot!(full_name, choices);
}

mod ristretto {
    use super::*;

    #[test]
    fn ciphertext_snapshot() {
        test_ciphertext_snapshot::<Ristretto>();
    }

    #[test]
    fn zero_encryption_snapshot() {
        test_zero_encryption_snapshot::<Ristretto>();
    }

    #[test]
    fn bool_encryption_snapshot() {
        test_bool_encryption_snapshot::<Ristretto>();
    }

    #[test]
    fn range_encryption_snapshot() {
        test_range_encryption_snapshot::<Ristretto>();
    }

    #[test]
    fn encrypted_choice_snapshot() {
        test_encrypted_choice_snapshot::<Ristretto>();
    }

    #[test]
    fn encrypted_multi_choice_snapshot() {
        test_encrypted_multi_choice_snapshot::<Ristretto>();
    }
}

mod k256 {
    use super::*;

    type K256 = Generic<::k256::Secp256k1>;

    #[test]
    fn ciphertext_snapshot() {
        test_ciphertext_snapshot::<K256>();
    }

    #[test]
    fn zero_encryption_snapshot() {
        test_zero_encryption_snapshot::<K256>();
    }

    #[test]
    fn bool_encryption_snapshot() {
        test_bool_encryption_snapshot::<K256>();
    }

    #[test]
    fn range_encryption_snapshot() {
        test_range_encryption_snapshot::<K256>();
    }

    #[test]
    fn encrypted_choice_snapshot() {
        test_encrypted_choice_snapshot::<K256>();
    }

    #[test]
    fn encrypted_multi_choice_snapshot() {
        test_encrypted_multi_choice_snapshot::<K256>();
    }
}
