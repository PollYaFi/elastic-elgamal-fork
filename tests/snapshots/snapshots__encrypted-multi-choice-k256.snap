---
source: tests/snapshots.rs
expression: choices

---
variants:
  - random_element: AgQbzI7_NulX5AZav0Ty90yRPsu9APHrKvYGo03s3Dn_
    blinded_element: Ap9uGRrxTZpokI5bSjq33H5WMsM6EwjF1o--egduVmgM
  - random_element: ArTsnXyCGVQRsyWMha-strO2-d3lZjFcnhysFN6qdz5F
    blinded_element: AvKO9hlU1CAgB1ti_cgIcKNbFknJhrGBQ-Jreq1TnPHs
  - random_element: A1vpeX89czeuNEIMmtycQuD-vyOBOVhf_k8lFjDpGS1x
    blinded_element: AudhDxb2dIyFKV70o_e0XUoYVWdycbYh9kY7c3yi-cWB
  - random_element: Ahu1rSSRo-gjzP35fyMKcrz0ZKtKoMi2VpC2rAK50v_M
    blinded_element: A_r6e0woor7xmm-vbFFE-r5bqP0h6BVJm5LWRHF0ZHwY
  - random_element: A69jbhMtuT9nhecpopt7jnG62L9Rlq6Gaai6t3oW4hFK
    blinded_element: AzhQRLqUpwIDs5jDvLVgb2XsKy68IzW0-0F1F6vyR1OJ
range_proof:
  common_challenge: H6eBWsmxZ5pusnsq38lgzjWoP7L78-I382BXmjpdHPM
  ring_responses:
    - 8x6rLuF6jjeb6TEl7rgbEfGMiGh3zU-R7OgxtHKEN94
    - A5GmkLDdIK_ElgFwnTy7KW9tnQxDslStKdohL-Lw-jo
    - dvzLv2IkNfu_cU_uzU5T1tr48nv-NKYqEzO-GFnhyKY
    - MYSBRZ9EDoNJ6UYDYdAsb5F2yoX7XZGOS7dvEGJ6RQs
    - AtxlAGMPfUxBlUGZd1QEkqsdVx9foXiJEJ2RkN0uY3U
    - i9CRAZDbXHwZFa_0mue9wY_o5yhFSmv9nMx1_aOJzxQ
    - 1HFSD8iyRxjDWIUPVKCtBhELA0HQjkOpduN8ayLTNlU
    - Hq4ttiTnxW5k1afBQG3TVK1aTiVAsQKvK348xq_jtJY
    - Rc6YO1oSKrOVLCUJsDvVJlA-isUazfUlNwwnvRPEFAI
    - O7Yjqso4805YQ06JAQdPShIHpqm8BHs6pLYntAOiO64
sum_proof: ~

